File: ts/src/integrations/browser/policy.ts
import {
  BROWSER_CONTRACT_SCHEMA_VERSION,
  validateBrowserSessionConfig,
  type BrowserAction,
  type BrowserPolicyDecision,
  type BrowserSessionConfig,
} from "./contract/index.js";
import type { BrowserSettingsLike } from "./types.js";

const INTERNAL_ALLOWED_URLS = new Set(["about:blank"]);

export function normalizeBrowserAllowedDomains(input: string | readonly string[]): string[] {
  let raw: string[];
  if (typeof input === "string") {
    raw = input.split(",");
  } else {
    raw = [...input];
  }
  const normalized: string[] = [];
  const seen = new Set<string>();
  for (const item of raw) {
    const domain = item.trim().toLowerCase();
    if (!domain || seen.has(domain)) {
      continue;
    }
    seen.add(domain);
    normalized.push(domain);
  }
  return normalized;
}

export function buildDefaultBrowserSessionConfig(
  overrides: Partial<BrowserSessionConfig> = {},
): BrowserSessionConfig {
  const config: BrowserSessionConfig = {
    schemaVersion: BROWSER_CONTRACT_SCHEMA_VERSION,
    profileMode: "ephemeral",
    allowedDomains: [],
    allowAuth: false,
    allowUploads: false,
    allowDownloads: false,
    captureScreenshots: true,
    headless: true,
    downloadsRoot: null,
    uploadsRoot: null,
    ...overrides,
  };
  return assertValidBrowserSessionConfig(config);
}

export function resolveBrowserSessionConfig(settings: BrowserSettingsLike): BrowserSessionConfig {
  return buildDefaultBrowserSessionConfig({
    profileMode: settings.browserProfileMode,
    allowedDomains: normalizeBrowserAllowedDomains(settings.browserAllowedDomains),
    allowAuth: settings.browserAllowAuth,
    allowUploads: settings.browserAllowUploads,
    allowDownloads: settings.browserAllowDownloads,
    captureScreenshots: settings.browserCaptureScreenshots,
    headless: settings.browserHeadless,
    downloadsRoot: settings.browserDownloadsRoot || null,
    uploadsRoot: settings.browserUploadsRoot || null,
  });
}

export function evaluateBrowserActionPolicy(
  config: BrowserSessionConfig,
  action: BrowserAction,
): BrowserPolicyDecision {
  if (action.type === "navigate") {
    return evaluateNavigationPolicy(config, action.params.url);
  }
  if (action.type === "fill" && action.params.fieldKind === "password" && !config.allowAuth) {
    return { allowed: false, reason: "auth_blocked" };
  }
  return { allowed: true, reason: "allowed" };
}

function assertValidBrowserSessionConfig(config: BrowserSessionConfig): BrowserSessionConfig {
  const validation = validateBrowserSessionConfig(config);
  if (!validation.valid) {
    throw new TypeError(`invalid browser session config: ${validation.errors.join("; ")}`);
  }
  return config;
}

function evaluateNavigationPolicy(
  config: BrowserSessionConfig,
  url: string,
): BrowserPolicyDecision {
  if (INTERNAL_ALLOWED_URLS.has(url)) {
    return { allowed: true, reason: "allowed" };
  }

  const parsed = parseNavigationTarget(url);
  if (!parsed.valid) {
    return { allowed: false, reason: "invalid_url" };
  }
  if (parsed.inlineCredentials && !config.allowAuth) {
    return { allowed: false, reason: "auth_blocked" };
  }

  for (const allowedDomain of config.allowedDomains) {
    if (matchesAllowedDomain(parsed.hostname, allowedDomain)) {
      return { allowed: true, reason: "allowed", matchedDomain: allowedDomain };
    }
  }
  return { allowed: false, reason: "domain_not_allowed" };
}

function parseNavigationTarget(url: string): {
  readonly valid: true;
  readonly hostname: string;
  readonly inlineCredentials: boolean;
} | {
  readonly valid: false;
} {
  try {
    const parsed = new URL(url);
    if (parsed.protocol !== "http:" && parsed.protocol !== "https:") {
      return { valid: false };
    }
    if (!parsed.hostname) {
      return { valid: false };
    }
    return {
      valid: true,
      hostname: parsed.hostname.toLowerCase(),
      inlineCredentials: parsed.username.length > 0 || parsed.password.length > 0,
    };
  } catch {
    return { valid: false };
  }
}

function matchesAllowedDomain(hostname: string, allowedDomain: string): boolean {
  if (allowedDomain.startsWith("*.")) {
    const suffix = allowedDomain.slice(2);
    return hostname.length > suffix.length && hostname.endsWith(`.${suffix}`);
  }
  return hostname === allowedDomain;
}

File: autocontext/src/autocontext/integrations/browser/policy.py
"""Pure policy helpers for browser exploration."""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Any
from urllib.parse import urlparse

from autocontext.integrations.browser.contract.types import BrowserAction, BrowserSessionConfig
from autocontext.integrations.browser.validate import validate_browser_action, validate_browser_session_config

if TYPE_CHECKING:
    from autocontext.config.settings import AppSettings

INTERNAL_ALLOWED_URLS = {"about:blank"}


@dataclass(frozen=True, slots=True)
class BrowserPolicyDecision:
    allowed: bool
    reason: str
    matched_domain: str | None = None


def normalize_browser_allowed_domains(domains: str | list[str]) -> list[str]:
    raw = domains if isinstance(domains, list) else domains.split(",")
    normalized: list[str] = []
    seen: set[str] = set()
    for item in raw:
        domain = item.strip().lower()
        if not domain or domain in seen:
            continue
        seen.add(domain)
        normalized.append(domain)
    return normalized


def build_default_browser_session_config(
    *,
    allowed_domains: list[str] | None = None,
    profile_mode: str = "ephemeral",
    allow_auth: bool = False,
    allow_uploads: bool = False,
    allow_downloads: bool = False,
    capture_screenshots: bool = True,
    headless: bool = True,
    downloads_root: str | None = None,
    uploads_root: str | None = None,
) -> BrowserSessionConfig:
    return validate_browser_session_config({
        "schemaVersion": "1.0",
        "profileMode": profile_mode,
        "allowedDomains": allowed_domains or [],
        "allowAuth": allow_auth,
        "allowUploads": allow_uploads,
        "allowDownloads": allow_downloads,
        "captureScreenshots": capture_screenshots,
        "headless": headless,
        "downloadsRoot": downloads_root,
        "uploadsRoot": uploads_root,
    })


def resolve_browser_session_config(settings: AppSettings) -> BrowserSessionConfig:
    return build_default_browser_session_config(
        allowed_domains=normalize_browser_allowed_domains(settings.browser_allowed_domains),
        profile_mode=settings.browser_profile_mode,
        allow_auth=settings.browser_allow_auth,
        allow_uploads=settings.browser_allow_uploads,
        allow_downloads=settings.browser_allow_downloads,
        capture_screenshots=settings.browser_capture_screenshots,
        headless=settings.browser_headless,
        downloads_root=settings.browser_downloads_root or None,
        uploads_root=settings.browser_uploads_root or None,
    )


def evaluate_browser_action_policy(
    config: BrowserSessionConfig,
    action: BrowserAction | dict[str, Any],
) -> BrowserPolicyDecision:
    parsed_action = validate_browser_action(action) if isinstance(action, dict) else action
    if parsed_action.type == "navigate":
        url = str(parsed_action.params.url)
        if url in INTERNAL_ALLOWED_URLS:
            return BrowserPolicyDecision(allowed=True, reason="allowed")
        target = _parse_navigation_target(url)
        if target is None:
            return BrowserPolicyDecision(allowed=False, reason="invalid_url")
        hostname, inline_credentials = target
        if inline_credentials and not config.allowAuth:
            return BrowserPolicyDecision(allowed=False, reason="auth_blocked")
        for allowed_domain in config.allowedDomains:
            if _matches_allowed_domain(hostname, str(allowed_domain)):
                return BrowserPolicyDecision(
                    allowed=True,
                    reason="allowed",
                    matched_domain=str(allowed_domain),
                )
        return BrowserPolicyDecision(allowed=False, reason="domain_not_allowed")
    if parsed_action.type == "fill":
        field_kind = parsed_action.params.fieldKind
        if field_kind == "password" and not config.allowAuth:
            return BrowserPolicyDecision(allowed=False, reason="auth_blocked")
    return BrowserPolicyDecision(allowed=True, reason="allowed")


def _parse_navigation_target(url: str) -> tuple[str, bool] | None:
    try:
        parsed = urlparse(url)
    except ValueError:
        return None
    if parsed.scheme not in {"http", "https"} or not parsed.hostname:
        return None
    inline_credentials = bool(parsed.username or parsed.password)
    return parsed.hostname.lower(), inline_credentials


def _matches_allowed_domain(hostname: str, allowed_domain: str) -> bool:
    if allowed_domain.startswith("*."):
        suffix = allowed_domain[2:]
        return len(hostname) > len(suffix) and hostname.endswith(f".{suffix}")
    return hostname == allowed_domain
